### PR TITLE
Refactor using 'ctr' and 'stream-cipher' crates

### PR DESCRIPTION
**Pull request by tarcieri**
The `ctr` crate provides a generic abstraction for building `StreamCipher` types out of a `BlockCipher`: counter mode (CTR).

Using the `ctr` and `stream-cipher` crates, and going back to a more trait-driven API, we can eliminate many of the (needless and unwieldy) generic type parameters introduced in bb7e0d08 (which used the `block-modes` crate for the first time).

Fixing this was the big blocker for another release, so hopefully after this 0.4 will be nearly ready to release.

This commit deletes all of the FFI code rather than attempting to update it to the new API. It should likely be replaced by [cbindgen](https://github.com/eqrion/cbindgen), rather than handrolled. This will be addressed in a follow-up PR prior to a 0.4 release.

### CODE REVIEW

File: src/cipher.rs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    Ctr128,
+};
+use generic_array::GenericArray;
+use zeroize::Zeroize;
+
+/// Stream ciphers (i.e. AES-CTR) used in AES-SIV constructions
+// TODO: replace this with traits from the `stream-cipher` crate?
+pub trait Cipher {
+    /// Create a new CTR instance
+    fn new(key: &[u8]) -> Self;
```
Comment (spacekookie) [inline src/cipher.rs]: Does the `stream-cipher` crate have a way of more gracefully failing the init of a Cipher if, i.e. not the right key length was provided?

Comment (tarcieri) [reply src/cipher.rs]: Haha this whole trait is bogus, although this is "ok" in as much as this is supposed to be an internal API, but isn't. Mostly I was just trying to refactor to the point my code compiled and worked before doing a second pass.

The `stream-cipher` crate uses `GenericArray<u8>` here and thus supports type safety around key lengths:

https://docs.rs/stream-cipher/0.3.0/stream_cipher/trait.NewStreamCipher.html#tymethod.new

Comment (spacekookie) [reply src/cipher.rs]: Ah, I see :grin: How is a failure encoded then though? Does it just return a `Self` where the key-length is something invalid?

Comment (tarcieri) [reply src/cipher.rs]: In my bogus trait it panics on an `assert_eq!` below. With `stream-cipher`, all keys that can be expressed as a `GenericArray<u8, Self::KeySize>` at a given key size are valid, so there is never a failure case to deal with.

Comment (tarcieri) [reply src/cipher.rs]: Update: after another round of refactoring, this trait no longer exists, and previous usages have been replaced with the `NewStreamCipher` and `SyncStreamCipher` traits.

**Pull request merged**